### ISSUE TITLE
Simply concat colored boxes result in visible seams

### ISSUE THREAD

**Issue by fzyzcjy** [flutter/flutter#106485]
## Steps to Reproduce

1. Execute `flutter run` on the code sample, look at it
2. If interested, remove `RepaintBoundary`, run again, look at it

**Expected results:** 
Should see a whole green screen without any seam

**Actual results:** 
has visible seam

Remark: This is only a reproducible sample; in my real-world case, the seam is quite ugly and affects the overall display

The screenshot:

![image](https://user-images.githubusercontent.com/5236035/175304772-f45286dd-d667-46ec-b88d-f388fca222ad.png)

Even if remove the RepaintBoundary in code, still buggy:

![image](https://user-images.githubusercontent.com/5236035/175304879-4c988d95-ef97-4633-a8d1-ab80151bf9d5.png)

<details>
<summary>Code sample</summary>

```dart
import 'dart:io';
import 'dart:ui' as ui;

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_test/src/_matchers_io.dart'; // ignore: implementation_imports
import 'package:path/path.dart' as p;

void main() {
  final binding = TestWidgetsFlutterBinding.ensureInitialized();
  binding.window
    ..physicalSizeTestValue = const Size(41, 41)
    ..devicePixelRatioTestValue = 1;

  testWidgets('example', (tester) async {
    await tester.pumpWidget(
      MaterialApp(
        debugShowCheckedModeBanner: false,
        home: Container(
          color: Colors.white,
          child: Column(
            children: [
              for (var i = 0; i < 2; ++i)
                Row(
                  children: [
                    for (var j = 0; j < 2; ++j)
                      RepaintBoundary(
                        child: Container(
                          color: Colors.green,
                          width: 20.5,
                          height: 20.5,
                        ),
                      )
                  ],
                ),
            ],
          ),
        ),
      ),
    );

    await tester.debugWidgetTestSaveScreenshot(find.byType(MaterialApp), 'debug_screenshot');
  });
}

// helper function
extension on WidgetTester {
  Future<void> debugWidgetTestSaveScreenshot([Finder? finder, String stem = 'debug_screenshot']) async {
    await runAsync(() async {
      final image = await captureImage(element(finder ?? find.byType(MaterialApp)));
      final bytes = (await image.toByteData(format: ui.ImageByteFormat.png))!.buffer.asUint8List();
      final path = p.join((goldenFileComparator as LocalFileComparator).basedir.path, '$stem.png');
      debugPrint(
          'debugWidgetTestSaveScreenshot save to path=$path image.size=${image.width}x${image.height} byte.length=${bytes.length}');
      File(path).writeAsBytesSync(bytes);
    });
  }
}

```

</details>

<details>
  <summary>Logs</summary>

```
no log
```

```
no error
```

```
3.0.2
```

</details>

**Comment by huycozy** [comment]
@fzyzcjy 

> Simply concat colored boxes result in visible seams
> If interested, remove RepaintBoundary, run again, look at it
> Even if remove the RepaintBoundary in code, still buggy:

The problem you are describing is not really clear whether it is related to RepaintBoundary or not.
I also tried changing the parameter of `physicalSizeTestValue` and it worked as expected (see demo). 

<details>
<summary>Demo</summary>

https://user-images.githubusercontent.com/104349824/175524942-cdd2cc28-6b72-4b96-bcd5-2fdff35ccbc4.mp4
</details>

As far as I see, it seems you are asking for assistance on personal code. Please see https://flutter.dev/community for resources and ask questions like this.
You may also get some help if you post it on Stack Overflow and if you need help with your code, please see https://www.reddit.com/r/flutterhelp.
Closing, as this isn't an issue with Flutter itself. If you disagree, please write in the comments and I will reopen it.

Thank you!

**Comment by fzyzcjy** [comment]
> As far as I see, it seems you are asking for assistance on personal code.
If you disagree, please write in the comments and I will reopen it.

Thanks for your reply, but I disagree. This is unrelated to personal code, but indeed a fundamental bug of Flutter itself. The bug is (after seeing your reproduction and modifications), when you have non-integer sized colored rectangles, its boundary is incorrectly rendered.

**Comment by fzyzcjy** [comment]
@huycozy Hi, "If you disagree, please write in the comments and I will reopen it." - I have written in the comments but have not seen it reopened

**Comment by fzyzcjy** [comment]
Hi thanks for the reply.

> Based on your response and your issue's description, RepaintBoundary isn't related to the issue, so please leave it out of the description as well as the sample code to clarify the issue.

post edited, removing repaintboundary.

> And you also need to describe the issue so that it's related to physicalSizeTestValue, then we have enough relevant information to proceed further.

I guess not related to setting physicalSizeTestValue, but related to non-integer sized widgets

> Furthermore, I see there's an open issue related to physicalSizeTestValue https://github.com/flutter/flutter/issues/62451 that you may find relevant information.
If you see that issue is similar, I will close this issue as a duplicate, so please follow up on it for further updates.

Seems no imho

**Comment by huycozy** [comment]
@fzyzcjy Thanks for updating the issue! Labeling the issue for further investigation!

**Comment by goderbauer** [comment]
@fzyzcjy Does this only show up in a test? Or can you reproduce that when running on an actual device?

**Comment by fzyzcjy** [comment]
@goderbauer This reproducible sample comes from some real-world cases in my app which runs on actual device

**Comment by fzyzcjy** [comment]
https://medium.com/flutter/whats-new-in-flutter-3-3-893c7b9af1ff

> Fractional translation
Previously, the Flutter Engine would always align composited layers to exact pixel boundaries because it improved rendering performance on older (32-bit) model iPhones. Since adding desktop support, we’ve noticed that this led to observable snapping behavior, since screen device pixel ratios are commonly much lower. For example, on low DPR screens, tooltips could be seen visibly snapping when fading in. After determining that this pixel snapping was no longer necessary for performance on newer iPhone models, [we’ve removed this pixel snapping](https://github.com/flutter/flutter/issues/103909) from the Flutter Engine to improve desktop fidelity. Additionally, we also found that removing this pixel snapping stabilized a number of our golden image tests, which would frequently change with small hairline rendering differences.

Not tested yet, but hope this is fixed in 3.3 :)
